src/components/modal-flow/ModalFlow.js

```diff
@@ -0,0 +1,373 @@
+    HALF_BORDERED_TO_WIDE_ENTERING: 'half-bordered-to-wide-entering',
+    WIDE_TO_FULL_EXITING: 'wide-to-full-exiting',
+    WIDE_TO_FULL_ENTERING: 'wide-to-full-entering',
+    FULL_TO_HALF_EXITING: 'full-to-half-exiting',
+    FULL_TO_HALF_ENTERING: 'full-to-half-entering',
+    FULL_TO_WIDE_EXITING: 'full-to-wide-exiting',
+    FULL_TO_WIDE_ENTERING: 'full-to-wide-entering',
+};
+
+class ModalFlow extends Component {
+    static getDerivedStateFromProps(props, state) {
+        if (state.variant && state.variant !== props.variant) {
+            console.error('The ModalFlow variant prop can\'t be changed');
+        }
+
+        const children = castArray(props.children);
+        const steps = children.map((child) => ({
+            id: child.props.id,
+        }));
+
+        if (state.steps !== null && !isEqual(steps, state.steps)) {
+            const subset = steps.slice(0, state.steps.length);
+
+            !isEqual(subset, state.steps) && console.error('New steps can only be added at the end');
+        }
+
+        const currentStepIndex = steps.findIndex((step) => step.id === props.step);
+        const currentStep = {
+            isFirst: currentStepIndex === 0,
+            isLast: currentStepIndex + 1 === steps.length,
+        };
+
+        // It means that the step index has changed
+        const requestNextStepIndex = currentStepIndex !== state.currentStepIndex && currentStepIndex;
+
+        const layout = ModalFlow.inferLayout(currentStep.isFirst, currentStep.isLast, props.variant);
+        const requestNextLayout = layout !== state.layout && layout;
+
+        if (layout === LAYOUT.HALF_BORDERED && state.layout === LAYOUT.WIDE) {
+            console.error(`This layout transition (${state.layout} to ${layout}) is not allowed`);
+        }
+
+        return {
+            variant: props.variant,
+            steps,
+            requestNextStepIndex,
+            requestNextLayout,
+        };
+    }
+
+    static inferLayout = (isFirstStep, isLastStep, variant) => {
+        switch (variant) {
+        case 'simple': return LAYOUT.HALF;
+        case 'simple-with-feedback': return isLastStep ? LAYOUT.FULL : LAYOUT.HALF;
+        case 'advanced':
+            if (isFirstStep) {
+                return LAYOUT.HALF_BORDERED;
+            } else if (isLastStep) {
+                return LAYOUT.FULL;
+            }
+
+            return LAYOUT.WIDE;
+
+        default: return LAYOUT.HALF;
+        }
+    };
+
+    prevLayout = null;
+    isAnimatingLayout = false;
+    layoutTransition = null;
+    stepsPlacement = 'right';
+
+    state = {
+        layout: null,
+        currentStepIndex: 0,
+        variant: null,
+        steps: null,
+        requestNextStepIndex: false,
+        requestNextLayout: false,
+    };
+
+    render() {
+        const { layout, requestNextLayout, requestNextStepIndex, variant } = this.state;
+
+        this.layoutTransition = this.inferLayoutTransition();
```

Comment:
Seems strange to be setting a class variable in the render method. Can this be done in another lifecycle method?

Reply:
I need to set this variable before each render. Doing this on `getDerivedStateFromProps` also seems weird to me. Furthermore, to use `inferLayoutTransition` method on `getDerivedStateFromProps` I need to change it to be a static method. That change wouldn't allow me to access to `this` instance (and I need to access it so that I can set `this.isAnimatingLayout`).